### PR TITLE
Implement draggable annotations on MapView

### PR DESCRIPTION
**Pull request by browniefed**
I am unsure about how the API for this should be implemented or if I even did anything correct.

I chose not to do a separate functions for each state change, but didn't know if the int enum was fine to pass and let people deal with it or if I should likely map it to a string but didn't know the appropriate way to do that on the Objective-C side, or do something else.
Intended to solve #2512

### CODE REVIEW

File: React/Modules/RCTPointAnnotation.h [hunk 1]
```diff
@@ -15,5 +15,7 @@
 @property (nonatomic, assign) BOOL hasLeftCallout;
 @property (nonatomic, assign) BOOL hasRightCallout;
 @property (nonatomic, assign) BOOL animateDrop;
+@property (nonatomic, assign) BOOL draggable;
+@property(nonatomic) MKAnnotationViewDragState dragState;
```
Comment (nicklockwood) [inline React/Modules/RCTPointAnnotation.h]: nit: space between @property and (nonatomic)

File: Libraries/Components/MapView/MapView.js [hunk 2]
```diff
@@ -252,8 +260,12 @@ var MapView = React.createClass({
     }
   },
 
+  _onAnnotationDragChange: function(event: Event) {
+    this.props.onAnnotationDragChange && this.props.onAnnotationDragChange(event);
+  },
+
   render: function() {
-    return <RCTMap {...this.props} onPress={this._onPress} onChange={this._onChange} />;
+    return <RCTMap {...this.props} onPress={this._onPress} onChange={this._onChange} onAnnotationDragChange={this._onAnnotationDragChange} />;
```
Comment (nicklockwood) [inline Libraries/Components/MapView/MapView.js]: This is long enough now that we should wrap the props:

    return (
      <RCTMap
        {...this.props}
        onPress={this._onPress}
        onChange={this._onChange}
        onAnnotationDragChange={this._onAnnotationDragChange}
      />
    );

File: React/Modules/RCTPointAnnotation.h (hunk 1 above)
Comment (nicklockwood) [inline React/Modules/RCTPointAnnotation.h]: Actually, is this even used? It doesn't seem like it's being set or read anywhere.

File: React/Modules/RCTPointAnnotation.m [hunk 3]
```diff
@@ -11,4 +11,6 @@
 
 @implementation RCTPointAnnotation
 
+@synthesize coordinate;
```
Comment (nicklockwood) [inline React/Modules/RCTPointAnnotation.m]: Why is this needed?

File: React/Modules/RCTPointAnnotation.h (hunk 1 above)
Comment (browniefed) [inline React/Modules/RCTPointAnnotation.h]: No it isn't, I'll remove it.

File: React/Views/RCTMapManager.m [hunk 4]
```diff
@@ -73,6 +74,27 @@ - (void)mapView:(RCTMap *)mapView didSelectAnnotationView:(MKAnnotationView *)vi
...
 
+- (void) mapView:(RCTMap *)mapView annotationView:(MKAnnotationView *)view didChangeDragState:(MKAnnotationViewDragState)newState fromOldState:(MKAnnotationViewDragState)oldState
+{
+  if ([view.annotation isKindOfClass:[RCTPointAnnotation class]]) {
+    
+    RCTPointAnnotation *annotation = (RCTPointAnnotation *)view.annotation;
+    
+    if (mapView.onAnnotationDragChange) {
+      mapView.onAnnotationDragChange(@{
+        @"state": @(newState),
```
Comment (nicklockwood) [inline React/Views/RCTMapManager.m]: Is this state property useful? If it is, we should probably convert it to a more JS-friendly string constant, and define an enum for it in the js module.

**Comment by nicklockwood** [comment]
This looks pretty good. A few minor issues to sort out before it's shippable though.

File: React/Views/RCTMapManager.m (hunk 4 above)
Comment (browniefed) [inline React/Views/RCTMapManager.m]: The state is a enum that indicate state of the drag "dragging", "canceled" , "drop", etc. I'll get right on this.

File: React/Modules/RCTPointAnnotation.m (hunk 3 above)
Comment (browniefed) [inline React/Modules/RCTPointAnnotation.m]: I don't think it is. Someone online in my googlings had said it was necessary to auto generate the getters and setters, but I will remove it.

**Comment by browniefed** [comment]
Updated with changes and recommendations.

**Comment by nicklockwood** [comment]
lgtm @facebook-github-bot shipit

**Comment by chris-erickson** [comment]
Is this stuck open?

**Comment by nicklockwood** [comment]
@chris-erickson no, it never landed. I think there was a conflict. I'll investigate.

**Comment by browniefed** [comment]
I may have destroyed the branch in switching computers and git branches and such. If need be I can get a new PR opened up, and see about taking care of the conflicts.

**Comment by browniefed** [comment]
@nicklockwood new PR is here for your review https://github.com/facebook/react-native/pull/4441. Matched the new style in the JS file of creating a function variable vs `this.function` and renamed stuff for `RCTMapAnnotation` instead of `RCTPointAnnotaiton`.
